### PR TITLE
Comprehensive Codebase Refactor: DTO Updates, Database Migration Enhancements, Middleware Improvements, and Model Adjustments

### PR DESCRIPTION
**Pull request by atha3417**
This pull request includes several significant changes across multiple files, primarily focusing on refactoring to use Data Transfer Objects (DTOs), improving error handling, and updating database migration functions. The most important changes are summarized below:

### Refactoring to use DTOs:
* [`controllers/auth_controller.go`](diffhunk://#diff-d9f51a1c922e9ecffd0f70e862ab0c1bd5b44ac66d8c8f5950baeec901264705R7-R21): Replaced `models.User` with `dto.UserRequest` for request validation and parsing. [[1]](diffhunk://#diff-d9f51a1c922e9ecffd0f70e862ab0c1bd5b44ac66d8c8f5950baeec901264705R7-R21) [[2]](diffhunk://#diff-d9f51a1c922e9ecffd0f70e862ab0c1bd5b44ac66d8c8f5950baeec901264705L32-R43)
* [`controllers/category_controller.go`](diffhunk://#diff-a74b68b30ec7ea0be58f8cd6a158233b7796b608f0e882515776523c82421197R6-R79): Replaced `models.Category` with `dto.CategoryRequest` and `dto.CategoryResponse` for request validation and response formatting.
* [`controllers/menu_controller.go`](diffhunk://#diff-e19457272ccb13a5f448a97da761ef76d6f44363f305191ad43031e26a69439fL4-R6): Replaced `models.Menu` with `dto.MenuRequest` and `dto.MenuResponse` for request validation and response formatting. [[1]](diffhunk://#diff-e19457272ccb13a5f448a97da761ef76d6f44363f305191ad43031e26a69439fL4-R6) [[2]](diffhunk://#diff-e19457272ccb13a5f448a97da761ef76d6f44363f305191ad43031e26a69439fL19-R31) [[3]](diffhunk://#diff-e19457272ccb13a5f448a97da761ef76d6f44363f305191ad43031e26a69439fL31-R104)

### Error handling improvements:
* [`controllers/auth_controller.go`](diffhunk://#diff-d9f51a1c922e9ecffd0f70e862ab0c1bd5b44ac66d8c8f5950baeec901264705L32-R43): Improved error handling in the `Login` function to return more specific error messages.
* [`middlewares/require_admin_middleware.go`](diffhunk://#diff-41cf1307a3c1710d4be326daa507d9e0bbdf696e685f81af196c006bf93dffeeL6-R20): Simplified token parsing and improved error handling for unauthorized access.
* [`middlewares/require_auth_middleware.go`](diffhunk://#diff-2aba9a69ab7d1786ed3f24b1803cce4c71c80873ee06627a1e79f6681ba49f43L22-R25): Enhanced token validation and error handling for authentication middleware. [[1]](diffhunk://#diff-2aba9a69ab7d1786ed3f24b1803cce4c71c80873ee06627a1e79f6681ba49f43L22-R25) [[2]](diffhunk://#diff-2aba9a69ab7d1786ed3f24b1803cce4c71c80873ee06627a1e79f6681ba49f43R46-R48)

### Database migration updates:
* [`database/migration.go`](diffhunk://#diff-5692f665a628e9ac2654d8d371a1a09b1929d8830bc1b3d20a15ba08023ec4ddL8-R20): Split the `Migrate` function into `MigrateUp` and `MigrateDown` functions for better control over database schema changes.
* `main.go` and `api/index.go`: Updated to call `MigrateUp` and `MigrateDown` functions for database initialization and cleanup. [[1]](diffhunk://#diff-2873f79a86c0d8b3335cd7731b0ecf7dd4301eb19a82ef7a1cba7589b5252261L21-R23) [[2]](diffhunk://#diff-18ca9ac0c83e1edbd6e4f8881b0991a18cab9a62897b4214c3017f0085170763L25-R27)

### Model updates:
* `models/category.go`, `models/menu.go`, `models/user.go`: Refactored models to remove JSON and form tags, ensuring proper database constraints and relationships. [[1]](diffhunk://#diff-feefc5018db042024992c97f1b37278e6a62746f0825f4e10fefd78a9ab24678L4-R6) [[2]](diffhunk://#diff-6b9ee48fcd6016d9465a0721ea94dd3ce2183113fc4c4ec49f033f629e9da806L4-R10) [[3]](diffhunk://#diff-f9c27c5d201f4eb9a23b855d52131aab43a1842aeedcf9356797c4ae75c224dfR3-R22)

### Repository updates:
* [`repositories/category_repository.go`](diffhunk://#diff-c268101557b63ec0098289b8a6298ef92e9b3e06383b6640058d2819d09fbaf3L15-R19): Modified `GetCategory` to return the `Category` model and error, instead of just an error.
* [`repositories/menu_repository.go`](diffhunk://#diff-dd3902cf3c763747fb407005123ef9de9e88fc0ddbf43a845bc4743e48e80ea9L13-L25): Removed redundant code for setting `CategoryName` in `Menu` model.
* [`repositories/user_repository.go`](diffhunk://#diff-3b7066fe9a309fb747c0b28e7d520e71c740c016f71b49b6a0272c7c1bf323c1L8-R20): Added `CreateUser` function for creating new user records.

### CODE REVIEW

File: .env.example
Comment (ahargunyllib) [inline .env.example:1]: ini kenapa di ganti wkwkwkw, biasanya api 8000 atau ngga 8080 sih, 3000 buat fe. tapi gapapa

File: controllers/auth_controller.go [hunk 1]
```diff
@@ -29,15 +27,15 @@ func Login(c *fiber.Ctx) error {
 		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{"message": "Validation failed", "errors": errors})
 	}
 
-	userDB, _ := repo.GetUserByUsername(user.Username)
-	if userDB.ID == 0 {
+	userDB, err := repo.GetUserByUsername(userReq.Username)
+	if err != nil {
```
Comment (ahargunyllib) [inline controllers/auth_controller.go:31]: ini kalo error dari db better 500 internal server error sih, tp gapapa

File: dto/category_dto.go
Comment (ahargunyllib) [inline dto/category_dto.go:1]: ini better di pisah sih buat tiap usecase, cth: create sm edit

File: models/menu.go
Comment (ahargunyllib) [inline models/menu.go:1]: ini bukannya harus ada references nya ya? aman kah kek gini

File: models/user.go [hunk 2]
```diff
@@ -1,8 +1,28 @@
...
+
+const (
+	RoleUser  = "User"
+	RoleAdmin = "Admin"
+)
+
+func (u *User) BeforeCreate(tx *gorm.DB) (err error) {
+	u.ID = uuid.New()
+
+	if u.Username == "" || u.Password == "" {
```
Comment (ahargunyllib) [inline models/user.go:24]: harusnya ini ga perlu. keknya kalo insert empty string bakal dianggap null, jd buat aja username sm passwordnya not nullable

File: .env.example
Comment (atha3417) [reply .env.example:1]: OH YA, itu lupa soalnya kemarin lagi ada app yg make port 8000, jdnya make 3000 dulu, aku balikin

File: models/menu.go
Comment (atha3417) [reply models/menu.go:1]: image
aman karena udah ada nilai default

File: controllers/auth_controller.go [hunk 3]
```diff
@@ -29,6 +29,10 @@ func Login(c *fiber.Ctx) error {
 
 	userDB, err := repo.GetUserByUsername(userReq.Username)
 	if err != nil {
+		return c.Status(fiber.StatusUnauthorized).JSON(fiber.Map{"message": "Failed to login", "error": err})
```
Comment (ahargunyllib) [inline controllers/auth_controller.go:32]: ini harusnya ada if error is gorm.ErrNotFound baru invalid credentials, sisanya baru gini

File: controllers/menu_controller.go [hunk 4]
```diff
@@ -74,10 +73,10 @@ func CreateMenu(c *fiber.Ctx) error {
 		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{"message": "Validation failed", "errors": errors})
 	}
 
-	category := models.Category{ID: menuReq.CategoryID}
+	category, _ := repositories.GetCategory(int(menuReq.CategoryID))
```
Comment (ahargunyllib) [inline controllers/menu_controller.go:76]: ini kenapa di cast

**Review by ahargunyllib — CHANGES_REQUESTED** [review]

File: controllers/menu_controller.go (hunk 4 above)
Comment (atha3417) [reply controllers/menu_controller.go]: karena kalo ga nanti error, menuReq.CategoryId tipe nya uint soalnya

Comment (atha3417) [reply controllers/menu_controller.go]: soalnya menuReq.CategoryID tipenya uint

File: .env.example
[the remark this answers is not in the document]
Comment (atha3417) [reply .env.example:1]: biar sama semua pake double quote, ga cuma jwt secret sih

**Pull request merged**